### ISSUE TITLE
Suggesting auto fixes for violated parts to fix reports issued by rules

### ISSUE THREAD

**Issue by schalkms** [detekt/detekt#2560]
Suggesting code fixes for violated parts in the code base could be a nice add-on for detekt and its users/integrators.
I know that we decided (with good reasoning) to not mess with user code. Anyway, this is a different story, since this feature doesn't mess with source code without any user intervention.

It's nice for 1st time contributors to exactly notice, what has triggered a violation and how to fix the issue.
Since this allows to fix issues automatically with the help of GitHubs auto suggestion feature, for instance, this could lead to new plugins or use cases for detekt's users and integrators.
Auto fixes are one of the most awesome things for code reviews and saves time for the human reviewer.

For instance, the following rule could be a good starting point.
**issue: LineEndsWithWhitespace**
```diff
- println("Hello World") 
+ println("Hello World")
```

**Comment by BraisGabin** [comment]
Thinking about detekt, which is the difference between suggesting auto fixes and auto fix the code? I mean, which is the different in the output of detekt? Instead of change the code create `diff` files that you could apply to the code if you like them?

After reading your suggestion in the other PR I hacked around a bit and I wrote this: https://github.com/BraisGabin/seshat it is just a prove of concept.

This is a backend that receives a diff file and transforms it to comments in a PR as code sugestions. And it works: https://github.com/BraisGabin/ktlint-bot-playground/pull/1 (the comment was writted by the code in seshat)

The idea:
- Run your code formatter/fixer in your CI
- `git diff > information.diff`
- Upload the `information.diff` to a github app
- The github app parses the file and comment in the PR the changes done by the code formatter/fixer

The good part of this is that it works for any tool you want and any language. This is just a prove of concept I don't know if I'll continue with it.

**Comment by cortinico** [comment]
> This is just a prove of concept I don't know if I'll continue with it.

I really like the idea. I was also looking for something like that and I haven't found much around. Happy to see that there is development on this side 👍 Happy to support making this a production ready tool.

**Comment by schalkms** [comment]
To bring this in a production ready state I think detekt has to assert the reported locations in the code by our rules more thoroughly.  I once formulated this in issue #2087.
@cortinico thanks btw for your awesome support. 🙂

**Comment by BraisGabin** [comment]
> An incubation in the detekt group could be an option to make this tool more visible and attract some contributors similar to the `detekt-compiler-plugin`.

Does the project fit in detekt? I mean, the visibility that detekt can give is great, but I think that the goals are different. Just as an example, right now I'm using ktlint as the first integration. detetk doesn't have a option to provide fixes (or suggestions) yet.

And the detekt focus is kotlin. This project is language agnostic. Any tool that edits text files could be integrated here.

> Happy to support making this a production ready tool.

We can write a roadmap and see how to make this production ready. PRs or similar have little sense right now because the project is a real mess. Single backend file. Bad naming. No architecture. Bad practices. 0 security... It needs to be refactored completely.

----

I worked a bit more in the project. Progress:

- Now the messages are sent by the Github App instead of my own account.
- I have a fully functional [workflow](https://github.com/BraisGabin/ktlint-bot-playground/blob/master/.github/workflows/seshat.yml) installed in [this repository](https://github.com/BraisGabin/ktlint-bot-playground). If you open a PR to that repo with a kotlin file that doesn't follow the ktlint guidelines you will get an automatic message with the suggestion to fix it.

But there are a lot of things missing:
- Error handling
- The code in the [backend project](https://github.com/BraisGabin/seshat) is a mess. Single file backend.
- I need to try the [annotations in the checks API](https://developer.github.com/v3/checks/runs/#annotations-object) they seem the perfect match. But I don't know if they support suggestions.
- Right now it's kind of difficult to intergrate this in a repo: You need to install de Github App and you need to create the workflow. It would be nice to have a web console where you can decide in which repo you want to use this tool. Then the console will ensure that you have the Github App installed and will create the workflow for you.

**Comment by BraisGabin** [comment]
> > Instead of change the code create diff files that you could apply to the code if you like them?
> 
> Do you mean that the user sees the original file and a copy, which contains the auto fix? Afterwards, the user can decide, whether the suggestion should be applied. The `diff` can also be uploaded to GitHub in order to show the proposed changes. Do I understand you correctly?

Yes, what I mean with this is: How can we do detekt "suggest" the changes? The `cli` should create some output. The easiest one is to edit the code with the actual "suffestion". This is nice because you have git to show you the differences. But, as you say, this has serious trust problems.

The other output that I can think of is create diff files. One per issue (or just one, or one per file). This way we don't touch the code. But the change can be applied if the user wants really easy: `patch < unusedPrivate.diff`. The problem is that review diff files is bit difficult if you don't have other tools. I think that we could expand our html report to have the suggestions too. And even a button to add to the clipboard the command that you need to run to autofix that issue.

Does this have sense? Any other idea about the output format for this? I think that the output should work without github but should be expandable to allow any other tool to use it. I don't know if there is any standar around this. Something like checkstyle for report issues.

**Comment by schalkms** [comment]
@BraisGabin do you still actively work on this project? I think it's a hidden gem, which should gain more visibility in the detekt ecosystem. 🙂

**Comment by BraisGabin** [comment]
No really. I still think about it but I didn't implement anything new in 8 months.

**Comment by cortinico** [comment]
Do we still want this? Should we add the "help wanted"?

**Comment by BraisGabin** [comment]
I think that we should talk about the "auto-correct" feature in detekt and how do we want it to work in 2.0

We know that it introduces issues in the core of detekt:

- If we fix first and the run the "normal" issues some rules could receive a "corrupted" KtFile
- If we fix at the and the reports aren't correct because the text position change

I still think that detekt should only provide patch files and let the user apply them. We could create a task to apply them all.

But even with that in mind, right now we have false positive and false negatives that we need to fix. If we add the "suggestions" this project will require even more maintenance and I'm not sure if we are able to provide that.

So I would vote to think if we want to change our curren "auto-correct" from modify directly files to generate `.patch` files. But I wouldn't implement that feature to any of the default rules that we provide for now.

**Comment by cortinico** [comment]
Adding to the milestone work for 2.x

**Comment by dimsuz** [comment]
> So I would vote to think if we want to change our curren "auto-correct" from modify directly files to generate .patch files. > But I wouldn't implement that feature to any of the default rules that we provide for now.

I really like this idea! I tried to think on how it could look if integrated into Detekt — I mean an actual user-visible API.

My thoughts were that to produce a `.patch` (`diff -u`) output detekt would need to know before/after state.

* "Before" state is known, in each `Rule` there is a `KtFile` which is passed for e.g. `preVisit`/`postVisit`
* "After" state has to be provided by the rule author, but how exactly?

Should this be a mapping from old AST elements to new ones like

```kotlin
class MyRule : Rule() {
  override fun buildAutoCorrectChangeSet(): Map<KtElement, KtElement> {...}
}
```
And then Detekt would apply these changes by iterating through `KtFile` elements replacing some of them using the provided mapping and then calling `ktFile.text()` on the source and the result and diffing?

But doing this with mapping will be a bit noisy and clumsy to specify I guess.
Maybe instead simply ask the rule author to provide a whole `KtFile` where they apply changes to any part of it they wish?

```kotlin
class MyRule : Rule() {
  override fun buildAutoCorrections(source: KtFile): KtFile  {...}
}
```

Won't this be hard too? Because this would split my work as a rule author: first I do some clever AST tricks to find errors and then in a completely separate function I must... do them again, but a bit different this time? Maybe it would be nicer to somehow append intended auto-corrections at places where I do my `report(...)` calls? But so far I didn't think much on how this could be done and how ergonomic this would be.

These are my rather raw thoughts, maybe you did discuss other approaches? I'd be interested to hear about them!

**Comment by schalkms** [comment]
@dimsuz No discussion regarding solution approaches for this issue has been started. If you are interested, you could look way back into the git history to investigate the approach back then, where *native* detekt rules were included in the default package.

**Comment by dimsuz** [comment]
> No discussion regarding solution approaches for this issue has been started

Actually my intention with the previous post was to try to start it :) 

I didn't know detekt already had an auto correction take prior to the last one. I'll try to investigate.

**Comment by BraisGabin** [comment]
The first option, the one that does both things at the same time is what ktlint does and I think that's a good way to do it. A boolean that decides if you should also do the change or not.

The difference here is that ktlint doesn't report issues if the autofix is provided and in our case I'm not sure if that would be true. Probably we will create the report and the fix. Both at the same time.

Also I'm not sure how the diffs should look like.
- One diff per issue
- one diff per rule
- one diff per run (all the fixes of all the rules applied at the same time)

**Comment by dimsuz** [comment]
Thanks, I'll study how ktlint does this (didn't think to look it up!).

> Also I'm not sure how the diffs should look like.

Personally I'd prefer a "one diff per run option", because my goal would be to do something like this on CI:

```
$ detekt --autofix > patch.diff
$ patch -i ./patch.diff
```

**Comment by schalkms** [comment]
> Thanks, I'll study how ktlint does this (didn't think to look it up!).

You could also take a look on how detekt implemented it in earlier versions. Note that detekt included auto correction natively for detekt rules. Back then, we didn't have a ktlint integration.

**Comment by schalkms** [comment]
In times of LLMs, I don't think that actively pursuing this feature is worth the investment, also considering the needed effort.
Therefore, I'll remove the 2.0 milestone and close this issue. If you have a different opinion, please provide a comment and I will reopen this issue.
